### PR TITLE
Instrument catalog with prometheus metrics

### PR DESCRIPTION
**Pull request by jhaals**
## Hey, I just made a Pull Request!

This PR adds Prometheus metrics to the catalog. Metrics are optional to enable.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by lowjoel** [comment]
Hey @jhaals - I was just discussing with a colleague these couple days that maybe we should hide this behind (yet another) interface. I know Prometheus is a pretty common standard, but just wondering if we want to directly mandate that we'd only emit Prometheus metrics.

Not sure what the API surface is going to look like in that case though. What's your knee-jerk reaction?

(For us it's not too much a problem, we use Datadog. But we were thinking about implementing metrics and figuring that directly writing Datadog-coupled code in Backstage would probably be a no-no.)

**Comment by jhaals** [comment]
Hi @lowjoel,

Great question, we initially planned to with OpenTelemetry to allow more exporters but the project is just too early in development to be able to use.
Creating a pluggable API surface in backstage would be a slippery slope and we hope to use library like OpenTelemetry to solve that for us. At least initially we are planning to keep metrics somewhat internal until we've gained enough confidence. We will definitely have a hard time supporting multiple providers, Prometheus is at this point of time the standard for CNCF projects. From my understanding the datadog-agent support scraping metrics endpoints. Another alternative could be to create your own InstrumentedProcessingEngine that simply delegates to the DefaultProcessingEngine

**Comment by lowjoel** [comment]
Yeah. I'm perfectly OK with using Prometheus since Datadog supports it. Just wondering what the thought process is behind. If Backstage is comfortable with prom-client then we might as well just anoint that as standard and make sure every plugin uses it.

File: plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts [hunk 1]
```diff
@@ -475,14 +479,8 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
...
           tx.client.config.client === 'sqlite3'
-            ? tx.raw(`datetime('now', ?)`, [
-                `${this.options.refreshIntervalSeconds} seconds`,
-              ])
-            : tx.raw(
-                `now() + interval '${Number(
-                  this.options.refreshIntervalSeconds,
-                )} seconds'`,
-              ),
+            ? tx.raw(`datetime('now', ?)`, [`${nextRefresh} seconds`])
```
Comment (jhaals) [inline plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts:482]: This should be deleted, used to test that smearing actually worked.

**Comment by Rugvip** [comment]
@lowjoel We just really want some metrics for the catalog to drive the development, and don't want to be blocked on creating a perfect or even good enough metrics solution. What we're adding here is highly experimental and will most likely be replaced in the future.

**Comment by lowjoel** [comment]
> @lowjoel We just really want some metrics for the catalog to drive the development, and don't want to be blocked on creating a perfect or even good enough metrics solution.

Perfectly reasonable.

> What we're adding here is highly experimental and will most likely be replaced in the future.

Then this leads to the question, would this be something that you'd provide an upgrade path for, or are you expecting deployments to not use this? We have had the exact same desire to know what's happening in our system, and we'd like to figure out the options available.

**Comment by Fox32** [comment]
Btw., very cool to see this. This is one topic that is blocking us from migrating to the new backend. This is quite similar to what we have internally for the old backend. Looking forward to this 🙂

**Comment by Rugvip** [comment]
@lowjoel It's likely that we'll make this at least a little bit stable so that we can have the community try it out and report back some results. Should be a good way of seeing how the catalog performs in various setups and iron out any issues. We do have some obvious stuff to deal with coming out of our internal testing first tho :grin:

File: packages/backend/src/index.ts [hunk 2]
```diff
@@ -124,6 +126,7 @@ async function main() {
   const service = createServiceBuilder(module)
     .loadConfig(config)
     .addRouter('', await healthcheck(healthcheckEnv))
+    .addRouter('', metricsHandler())
```
Comment (TejasQ) [inline packages/backend/src/index.ts:129]: I find this quite interesting! Pardon me if a silly question 😅 – how is `''` as a route different from

```suggestion
    .addRouter(metricsHandler())
```

?

Comment (freben) [reply packages/backend/src/index.ts:129]: This is not an Express router, it's our builder class. https://github.com/backstage/backstage/blob/6d7126c2a1db0a50393c87a88b05f5dcbcc9212f/packages/backend-common/src/service/types.ts?plain=1#L24

Comment (Rugvip) [reply packages/backend/src/index.ts:129]: Both args are required, so the first one compiles and the second one doesn't 😅. Now why the single arg version doesn't exist is a different question :p

Comment (TejasQ) [reply packages/backend/src/index.ts:129]: Ah, thanks for that both of y'all! I thought it implemented the Express API

**Comment by andrewthauer** [comment]
+1 for using OpenTelemetry eventually. There are already exporters for most major vendors and libraries (including Datadog and Prometheus). This means the OpenTelemetry protocol/libraries can be the abstractions themselves.

**Comment by freben** [comment]
One thing that we were really missing was a complete solution in opentelemetry corresponding to Histogram and Summary. If you have experience that helps us get on opentelemetry sooner rather than later, any and all input is welcome!

**Comment by andrewthauer** [comment]
Sorry, only started evaluating OT more deeply just recently. Haven't got to an actual proof of concept yet. However, I do see the JavaScript client library should have a metrics histogram aggregator. https://github.com/open-telemetry/opentelemetry-js/issues/927

**Comment by Fox32** [comment]
> One thing that we were really missing was a complete solution in opentelemetry corresponding to Histogram and Summary. If you have experience that helps us get on opentelemetry sooner rather than later, any and all input is welcome!

Yeah, we made similar experience. While we committed to go with opentelemetry in the long term internally, we have the issue that it just isn't delivering what we need yet. And it's hard to add that in the exporter to prometheus.

**Comment by andrewthauer** [comment]
Seems there are some loose ends around this part still in discussion - https://github.com/open-telemetry/opentelemetry-specification/issues/982

File: packages/backend/src/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * limitations under the License.
+ */
+
+import { useHotCleanup } from '@backstage/backend-common';
+import { RequestHandler, Request } from 'express';
+import promBundle from 'express-prom-bundle';
+import prom from 'prom-client';
+import * as url from 'url';
+
+const rootRegEx = new RegExp('^/([^/]*)/.*');
```
Comment (Rugvip) [inline packages/backend/src/metrics.ts:23]: I'd slap a big comment at the top here explaining the situation around how this is experimental, copying at your own risk, etc.

**Review by Rugvip — APPROVED** [review]
:shipit:

Could add a comment after release in order to get it shipped

File: packages/backend/src/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  return promBundle({
+    includeMethod: true,
+    includePath: true,
+    // Using includePath alone is problematic, as it will include path labels with high
+    // cardinality (e.g. path params). Instead we would have to template them. However, this
+    // is difficult, as every backend plugin might use different routes. Instead we only take
+    // the first directory of the path, to have at least an idea how each plugin performs:
+    normalizePath,
+    promClient: { collectDefaultMetrics: {} },
```
Comment (Fox32) [inline packages/backend/src/metrics.ts:61]: Btw, you are initializing the default metrics (like CPU and GC) twice here. Once explicitly and once in `promBundle`. But you are exporting them with different prefix names, e.g.: `backstage_nodejs_eventloop_lag_seconds` and `nodejs_eventloop_lag_seconds`.

Comment (freben) [reply packages/backend/src/metrics.ts:61]: Ohh good catch. We're only experimenting for now so no huge loss, but we'll take a look at that. One thing we've also pondering is, should we export with the `backstage_` prefix everywhere by force or not.

Comment (Fox32) [reply packages/backend/src/metrics.ts:61]: Internally we have prefixed all metrics that are backstage specific with `backstage_`, but we didn't touch the default nodejs metrics as it allows to reuse existing dashboards.

**Comment by awanlin** [comment]
Hi @wasadigi, as most of the metrics are experimental there hasn't been much documentation created around this from my experience. At one time this was integrated into the `backend-common` but then removed, thankfully they included instructions on how to add them back which might help you out: https://github.com/backstage/backstage/blob/master/packages/backend-common/CHANGELOG.md#040

**Pull request merged**